### PR TITLE
An attempt to create step-by-step howto

### PR DESCRIPTION
**Pull request by ytsarev**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
 ```sh
-helm repo add k8gb https://absaoss.github.io/k8gb/
-helm repo update
-helm -n k8gb upgrade -i k8gb k8gb/k8gb --create-namespace --wait
+make deploy-gslb-operator VALUES_YAML=~/k8gb/eu-cluster.yaml
+```
+
+* It should deploy k8gb pretty quickly. Let's check the pod status
+```sh
+ k -n k8gb get pod
```
Comment (donovanmuller) [inline README.md:100]:
```suggestion
 kubectl -n k8gb get pod
```

File: README.md [hunk 2]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+k8gb-76cc56b55-t779s                                       1/1     Running    0          39s
+k8gb-coredns-799984c646-qz88m                              1/1     Running    0          41s
+k8gb-etcd-operator-etcd-backup-operator-fccb8fdd-bx8v5     1/1     Running    0          40s
+k8gb-etcd-operator-etcd-operator-765f655575-pgvbh          1/1     Running    0          40s
+k8gb-etcd-operator-etcd-restore-operator-8977cd5c8-k8l7l   1/1     Running    0          40s
+```
+
+* As you can see one of the `etcd-cluster` pods is still in `Init` state. The embedded `etcd-operator` is automatically configuring isolated `etcd` cluster to act as a backend for `coredns` which is used as part of k8gb for coordinated DNS responses. Expected post installation result:
+```sh
+k -n k8gb get pod
```
Comment (donovanmuller) [inline README.md:114]:
```suggestion
kubectl -n k8gb get pod
```

File: README.md [hunk 3]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+k8gb-etcd-operator-etcd-restore-operator-8977cd5c8-k8l7l   1/1     Running   0          7m59s
+```
+
+* Deploy k8gb to second cluster by repeating the same steps with exception of
+  * Obviously switch context to 2nd cluster with e.g. `export KUBECONFIG=us-cluster`
+  * Create another custom `values.yaml` with `cp ~/k8gb/eu-cluster.yaml ~/k8gb/us-cluster.yaml`
+  * At this stage the only thing you need to make is a geo tag swap to enable cross cluster communication
+    * `clusterGeoTag` becomes `us`
+    * And `extGslbClustersGeoTags` becomes `eu`
+  * Run the installation pointing to new values file `make deploy-gslb-operator VALUES_YAML=~/k8gb/us-cluster.yaml`
```
Comment (donovanmuller) [inline README.md:133]:
```suggestion
* Deploy k8gb to the second cluster by repeating the same steps with the exception of:
  * Switch context to 2nd cluster with `export KUBECONFIG=us-cluster`
  * Create another custom `values.yaml` with `cp ~/k8gb/eu-cluster.yaml ~/k8gb/us-cluster.yaml`
  * Create another geo tag to enable cross cluster communication:
    * `clusterGeoTag` becomes `us`
    * `extGslbClustersGeoTags` becomes `eu`
  * Run the installation pointing to new values file `make deploy-gslb-operator VALUES_YAML=~/k8gb/us-cluster.yaml`
```

File: README.md [hunk 4]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+
+* When your 2nd cluster is visibly ready with `k -n k8gb get pod` check we can proceed with the sample application installation
```
Comment (donovanmuller) [inline README.md:135]:
```suggestion
* When your 2nd cluster is ready by checking with `kubectl -n k8gb get pod`, we can proceed with the sample application installation
```

File: README.md [hunk 5]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+
+* We will use well known testing community app of [podinfo](https://github.com/stefanprodan/podinfo)
+```sh
+helm repo add podinfo https://stefanprodan.github.io/podinfo
+k create test-gslb
```
Comment (donovanmuller) [inline README.md:140]:
```suggestion
kubectl create test-gslb
```

File: README.md [hunk 6]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+helm upgrade --install podinfo --namespace test-gslb --set ui.message="us" podinfo/podinfo
+```
+As you can see above we did set special geo tag message in podinfo configuratio matching cluster geo tag. It is just for demonstation purposes.
```
Comment (donovanmuller) [inline README.md:143]:
```suggestion
As you can see above we did set special geo tag message in podinfo configuration matching cluster geo tag. It is just for demonstration purposes.
```

File: README.md [hunk 7]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+
+* Check that podinfo is running
+```sh
+k -n test-gslb get pod
```
Comment (donovanmuller) [inline README.md:147]:
```suggestion
kubectl -n test-gslb get pod
```

File: README.md [hunk 8]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+NAME                       READY   STATUS    RESTARTS   AGE
+podinfo-5cfcdc9c45-jbg96   1/1     Running   0          2m18s
+```
+
+* Let's create Gslb CRD to enable global load balancing for this application. Notice the podinfo Service name
+```sh
+k -n test-gslb get svc
```
Comment (donovanmuller) [inline README.md:154]:
```suggestion
kubectl -n test-gslb get svc
```

File: README.md [hunk 9]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+NAME            TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)             AGE
+podinfo         ClusterIP   10.96.250.84    <none>        9898/TCP,9999/TCP   9m39s
+```
+
+* Create yaml `~/k8gb/podinfogslb.yaml` descring Gslb as in a sample below
```
Comment (donovanmuller) [inline README.md:159]:
```suggestion
* Create a custom resource `~/k8gb/podinfogslb.yaml` describing `Gslb` as per the sample below:
```

File: README.md [hunk 10]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+            path: /
+  strategy:
+    type: roundRobin # Use a round robin load balancing strategy, when deciding which downstream clusters to route clients too
 ```
 
-See [values.yaml](https://github.com/AbsaOSS/k8gb/blob/master/chart/k8gb/values.yaml)
-for customization options.
+* And apply the resource in the target app namespace
+```sh
+k -n test-gslb apply -f podinfogslb.yaml
```
Comment (donovanmuller) [inline README.md:182]:
```suggestion
kubectl -n test-gslb apply -f podinfogslb.yaml
```

File: README.md [hunk 11]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+gslb.k8gb.absa.oss/podinfo created
+```
+
+* Check Gslb resource
+```sh
+k -n test-gslb get gslb
```
Comment (donovanmuller) [inline README.md:188]:
```suggestion
kubectl -n test-gslb get gslb
```

File: README.md [hunk 12]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+NAME      AGE
+podinfo   39s
+```
+
+* Check Gslb resource status
+```sh
+k -n test-gslb describe gslb
```
Comment (donovanmuller) [inline README.md:195]:
```suggestion
kubectl -n test-gslb describe gslb
```

File: README.md [hunk 13]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+      172.17.0.7
+      172.17.0.8
+  Service Health:
+    podinfo.cloud.example.com:  Healthy
+Events:                         <none>
+```
+
+* In the output above you should see that Gslb detected the `Healthy` status of underlying `podinfo` standard Kubernetes Service
+
+* Check that internal k8gb dns servers are responding accordingly on this cluster
```
Comment (donovanmuller) [inline README.md:235]:
```suggestion
* Check that internal k8gb DNS servers are responding accordingly on this cluster
```

File: README.md [hunk 14]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+  * Pick one of the worker nodes to test with
+    ```sh
+    k get nodes -o wide
+    NAME                       STATUS   ROLES    AGE   VERSION   INTERNAL-IP   EXTERNAL-IP   OS-IMAGE       KERNEL-VERSION     CONTAINER-RUNTIME
+    test-gslb2-control-plane   Ready    master   53m   v1.17.0   172.17.0.9    <none>        Ubuntu 19.10   4.19.76-linuxkit   containerd://1.3.2
+    test-gslb2-worker          Ready    <none>   52m   v1.17.0   172.17.0.8    <none>        Ubuntu 19.10   4.19.76-linuxkit   containerd://1.3.2
+    test-gslb2-worker2         Ready    <none>   52m   v1.17.0   172.17.0.7    <none>        Ubuntu 19.10   4.19.76-linuxkit   containerd://1.3.2
+    test-gslb2-worker3         Ready    <none>   52m   v1.17.0   172.17.0.10   <none>        Ubuntu 19.10   4.19.76-linuxkit   containerd://1.3.2
+    ```
+  * Use `dig` to make a dns query to it
```
Comment (donovanmuller) [inline README.md:245]:
```suggestion
  * Use `dig` to make a DNS query to it
```

File: README.md [hunk 15]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+    ```sh
+    dig +short @172.17.0.10 podinfo.cloud.example.com
+    172.17.0.8
+    172.17.0.10
+    172.17.0.7
+    ```
+  * One of your workers should already return DNS responses constructed by Gslb basing on service healthines information
```
Comment (donovanmuller) [inline README.md:252]:
```suggestion
  * One of your workers should already return DNS responses constructed by Gslb based on service health information
```

File: README.md [hunk 16]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+  * If edgeDNS was configured you can query your standard infra DNS directly and it should return the same
+    ```sh
+    dig +short podinfo.cloud.example.com
+    172.17.0.8
+    172.17.0.10
+    172.17.0.7
+    ```
+* Now it's time to deploy this application to the first `eu` cluster. The steps and configuration are exactly the same. Just changing ui message to `eu`
```
Comment (donovanmuller) [inline README.md:260]:
```suggestion
* Now it's time to deploy this application to the first `eu` cluster. The steps and configuration are exactly the same. Just changing `ui.message` to `eu`
```

File: README.md [hunk 17]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+```sh
+k create test-gslb
```
Comment (donovanmuller) [inline README.md:262]:
```suggestion
kubectl create test-gslb
```

File: README.md [hunk 18]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+helm upgrade --install podinfo --namespace test-gslb --set ui.message="eu" podinfo/podinfo
+```
+
+* Apply exactly the same Gslb definition
+```sh
+k -n test-gslb apply -f podinfogslb.yaml
```
Comment (donovanmuller) [inline README.md:268]:
```suggestion
kubectl -n test-gslb apply -f podinfogslb.yaml
```

File: README.md [hunk 19]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+dig +short podinfo.cloud.example.com
+172.17.0.8
+172.17.0.5
+172.17.0.10
+172.17.0.7
+172.17.0.6
+172.17.0.3
+```
+
+* And for final end-to-end test we can actually http query the application
```
Comment (donovanmuller) [inline README.md:327]:
```suggestion
* And for the final end-to-end test, we can use `curl` to query the application
```

File: README.md [hunk 20]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+  "message": "\"us\"",
+
+curl -s podinfo.example.com|grep message
+  "message": "\"us\"",
+
+curl -s podinfo.example.com||grep message
+  "message": "\"eu\"",
+```
+
+* As you can see specially marked `podinfo` returns differnet geo tags showing us the Global Round Robin strategy is working as expected
```
Comment (donovanmuller) [inline README.md:342]:
```suggestion
* As you can see specially marked `podinfo` returns different geo tags showing us the Global Round Robin strategy is working as expected
```

File: README.md [hunk 21]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+
+Hope you enjoyed the ride!
+
+If anything unclear or is going wrong feel free to contact us at https://github.com/AbsaOSS/k8gb/issues. We will appreciate any feedback/bugreport and especially Pull Request.
```
Comment (donovanmuller) [inline README.md:346]:
```suggestion
If anything unclear or is going wrong feel free to contact us at https://github.com/AbsaOSS/k8gb/issues. We will appreciate any feedback/bug report and Pull Requests are welcome.
```

File: README.md [hunk 22]
```diff
@@ -46,18 +46,306 @@ k8gb is very well tested with the following environment options
...
+
+For more advanced technical stuff and fully automated abstracted local install see below.
```
Comment (donovanmuller) [inline README.md:348]:
```suggestion
For more advanced technical documentation and fully automated local installation steps, see below.
```

**Comment by ytsarev** [comment]
@donovanmuller many thanks for this very thorough review! I've applied all suggestions and squashed the commits

**Pull request merged**